### PR TITLE
Implementing logic

### PR DESCRIPTION
**Pull request by ggleszczynski**

### CODE REVIEW

File: bt_ddos_shield/miner_shield.py [hunk 1]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
-        Ban a validator by its hotkey. Function blocks execution until manifest file is updated and info about file
-        is published to Bittensor.
+        Ban a validator by its hotkey. Task will be executed by worker. It will update manifest file and publish info
+        about new file version to blockchain.
 
         Args:
             validator_hotkey: The hotkey of the validator.
         """
+        self._add_task(MinerShieldBanValidatorTask(validator_hotkey))
+        pass
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:108]:
```suggestion
```

File: bt_ddos_shield/miner_shield.py [hunk 2]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+
+    def _add_task(self, task):
+        """
+        Add task to task queue. It will be handled by _worker_function.
+        """
+        if not isinstance(task, MinerShieldTask):
+            raise Exception("Task is not instance of MinerShieldTask")
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:115]:
```suggestion
    def _add_task(self, task: MinerShieldTask):
        """
        Add task to task queue. It will be handled by _worker_function.
        """
```
Traditionally we don't implement type enforcement in the first lines of every method and function. Rather than running it in runtime, we have external tools for that which perform static analysis (whenever possible).

File: bt_ddos_shield/miner_shield.py [hunk 3]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+        if not self.run:
+            raise Exception("Shield is disabled")
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:117]: In order to raise any exceptions you'll need to
```
class ShieldException(Exception):
    pass


class ShieldIsDisabled(ShieldException):
    pass
```
and raise that - this way the user of the module will be able to catch the intended exceptions without catching every exception, which includes KeyError, ValueError etc

However, in this particular case you should raise a built-in `ProgrammingError`, which is a builtin for cases like that.

File: bt_ddos_shield/miner_shield.py [hunk 4]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+        Function called in separate thread by enable() to start the shield. It is handling events put to task_queue.
+        """
+
+        self.event_processor.add_event(MinerShieldEvent(f"Starting shield"))
+
+        while self.run:
+            task = self.task_queue.get()
+            try_count = 1
+
+            while self.run:
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:132]:
```suggestion
            try_count: int = 0

            while try_count <= self.options.max_retries:
```

File: bt_ddos_shield/miner_shield.py [hunk 5]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+                    sleep(self.options.retry_delay)
+
+        self.event_processor.add_event(MinerShieldEvent(f"Stopping shield"))
+
+    def _handle_initialize(self):
+        """
+        Initialize shield. Load state and initial validators set.
+        """
+        self.state_manager.get_state()
+        self.event_processor.add_event(MinerShieldEvent("State loaded"))
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:155]: Rather than seeing
```py
self.event_processor.add_event(MinerShieldEvent("State loaded"))
```
everywhere, I'd prefer to see
```py
self._event("State loaded")
```

with
```py
def _event(self, description: str, *args, **kwargs):
    return self.event_processor.add_event(self.event_class(description, *args, **kwargs))
```
and `event_class=MinerShieldEvent` in the constructor

File: bt_ddos_shield/miner_shield.py [hunk 6]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+
+        self.validators_manager.refresh_validators()
+        validators: dict[Hotkey, str] = self.validators_manager.get_validators()
+        self.event_processor.add_event(MinerShieldEvent(f"Validators refreshed, got {len(validators)} validators"))
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:159]:
```suggestion
        self._event("Validators refreshed", validator_count=len(validators)))
```

Comment (ppolewicz) [reply bt_ddos_shield/miner_shield.py:159]: the MinerShieldEvent class will need to get a little bit more complicated, but we'll be able to emit smart objects that can then be filtered, sorted etc

File: bt_ddos_shield/miner_shield.py [hunk 7]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+        """
+        Run task logic.
+
+        Args
+            miner_shield: Instance of MinerShield in which task is executed.
+        """
+        pass
+
+    def __repr__(self):
+        return self.task_name
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:241]:
```suggestion
    NAME_DELETER = re.compile(r'^MinerShield(.*)Task$')

    @abstractmethod
    def run(self, miner_shield: MinerShield):
        pass

    def __repr__(self):
        return self.NAME_DELETER.sub(r'\1', self.__class__.__name__)
```

File: bt_ddos_shield/miner_shield.py [hunk 8]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+
+    def _handle_ban_validator(self, validator_hotkey: Hotkey):
+        """
+        Ban validator by its hotkey. It will update manifest file and publish info about new file version to blockchain.
+        """
+        # TODO
         pass
+
+
+class MinerShieldTask:
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:217]:
```suggestion
class AbstractMinerShieldTask(ABC):
```

File: bt_ddos_shield/miner_shield.py [hunk 9]
```diff
@@ -45,30 +50,221 @@ def __init__(self, validators_manager: AbstractValidatorsManager, address_manage
...
+
+    def __repr__(self):
+        return self.task_name
+
+class MinerShieldInitializeTask(MinerShieldTask):
+    def __init__(self):
+        super().__init__("INITIALIZE_SHIELD")
+
+    def handle(self, miner_shield: MinerShield):
+        miner_shield._handle_initialize()
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:248]: This could be refactored away into the abstract class too, but it's not worth it - we wouldn't save much anyway and it'd confuse static code analyzers

File: bt_ddos_shield/miner_shield.py [hunk 10]
```diff
@@ -186,85 +208,120 @@ def _handle_validators_changed(self):
...
+        address = self.manifest_manager.add_mapping_file(self.state_manager.get_state().active_addresses)
+        self._event(f"Manifest updated, new address: {address}")
+        self._add_task(MinerShieldPublishManifestTask(address))
 
-class MinerShieldTask:
+    def _handle_publish_manifest(self, address: Address):
+        """
+        Publish info about current manifest file to blockchain.
+        """
+        current_address = Address.deserialize(self.blockchain_manager.get(self.miner_hotkey))
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:264]:
```suggestion
        current_address = self.ADDRESS_CLASS.deserialize(self.blockchain_manager.get(self.miner_hotkey))
```
```py
class costam:
    ADDRESS_CLASS: type = Address

class MyCostam:
    ADDRESS_CLASS: type = MyAddress

```

File: bt_ddos_shield/miner_shield.py [hunk 11]
```diff
@@ -186,85 +208,120 @@ def _handle_validators_changed(self):
...
+        self._event(f"Validator {validator_hotkey} added to banned set")
+        self._add_task(MinerShieldValidatorsChangedTask())
 
+    def _handle_update_manifest(self):
+        """
+        Update manifest file and schedule publishing it to blockchain.
+        """
+        address = self.manifest_manager.add_mapping_file(self.state_manager.get_state().active_addresses)
+        self._event(f"Manifest updated, new address: {address}")
+        self._add_task(MinerShieldPublishManifestTask(address))
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:258]: ...

File: bt_ddos_shield/miner_shield.py [hunk 12]
```diff
@@ -186,85 +208,120 @@ def _handle_validators_changed(self):
...
-
-class MinerShieldInitializeTask(MinerShieldTask):
-    def __init__(self):
-        super().__init__("INITIALIZE_SHIELD")
+        return self.NAME_DELETER.sub(r'\1', self.__class__.__name__)
 
-    def handle(self, miner_shield: MinerShield):
+class MinerShieldInitializeTask(AbstractMinerShieldTask):
+    def run(self, miner_shield: MinerShield):
         miner_shield._handle_initialize()
```
Comment (ppolewicz) [inline bt_ddos_shield/miner_shield.py:299]:
```suggestion


```

File: bt_ddos_shield/utils.py [hunk 13]
```diff
@@ -1 +1,4 @@
-Hotkey = str # type of Hotkey
+from typing import TypeAlias
+
+Hotkey: TypeAlias = str
+PublicKey: TypeAlias = str
```
Comment (ppolewicz) [inline bt_ddos_shield/utils.py:4]: :heart:

File: tests/test_miner_shield.py [hunk 14]
```diff
@@ -14,8 +14,8 @@ def test_start_stop(self):
         """
         Test if shield is properly starting and stopping.
         """
-        shield = MinerShield(None, None, None, None,
-                             None, LoggingMinerShieldEventProcessor(), MinerShieldOptions(retry_delay=1))
+        shield = MinerShield(None,None, None, None, None,
```
Comment (ppolewicz) [inline tests/test_miner_shield.py:17]:
```suggestion
        shield = MinerShield(None, None, None, None, None,
```

File: bt_ddos_shield/utils.py (hunk 13 above)
Comment (bswck) [reply bt_ddos_shield/utils.py:4]: If you develop at head and plan to use 3.12+, which I pretty much think _is_ true, according to https://github.com/bactensor/bt-ddos-shield/blob/c8e6b969a9612692053b6d991f5c1917e201dd7a/pyproject.toml#L8-L9

...then I suggest you use `type` statements from PEP 695 instead:

```suggestion
type Hotkey = str
type PublicKey = str
```

Read more:
- [PEP 695](https://peps.python.org/pep-0695/)
- [the `type` statement](https://docs.python.org/3/reference/simple_stmts.html#type)

File: bt_ddos_shield/miner_shield.py (hunk 10 above)
Comment (bswck) [reply bt_ddos_shield/miner_shield.py]: `ClassVar[type[Address]]` though?

**Comment by bswck** [comment]
See also #9 and #10. GLHF working on this!

File: bt_ddos_shield/event_processor.py [hunk 15]
```diff
@@ -12,17 +13,30 @@ class Event:
...
+    """
+    Event processor which logs events to console.
+    """
+
+    def add_event(self, event: MinerShieldEvent):
+        if event.exception is not None:
+            print(f"MinerShieldEvent: {event.event_description}\nException happened:")
+            print(traceback.format_exc())
+        else:
+            print(f"MinerShieldEvent: {event.event_description}")
```
[the remark this answers is not in the document]
Comment (bswck) [reply bt_ddos_shield/event_processor.py]: And if we ever print that, we'd rather use stderr.
`traceback.print_exc()` can be a sufficient routine for that.

File: bt_ddos_shield/miner_shield.py (hunk 10 above)
Comment (ggleszczynski) [reply bt_ddos_shield/miner_shield.py]: I've done it other, better way with usage of Serializer class

File: bt_ddos_shield/event_processor.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/event_processor.py]: I've changed it already in future commits. But now I'm printing exception captured inside MinerShieldEvent. Probably not, but there is possibility, that it wouldn't be last exception thrown and print_exc calls sys.exception() inside to get the last one.

File: bt_ddos_shield/utils.py (hunk 13 above)
Comment (ggleszczynski) [reply bt_ddos_shield/utils.py:4]: After this change, such constructs used in my code stopped working:
1. if isinstance(obj, Hotkey):
2. decoded_mapping[Hotkey(hotkey)]

So I'm not changing this, because I think TypeAlias is what I want here.

**Comment by ggleszczynski** [comment]
Checked it. And thanks :-)

**Comment by ggleszczynski** [comment]
> See also #9 and #10. GLHF working on this!

Checked it. And thanks :-)

File: bt_ddos_shield/miner_shield.py [hunk 16]
```diff
@@ -1,74 +1,481 @@
...
+        return self.NAME_DELETER.sub(r'\1', self.__class__.__name__)
+
+
+class MinerShieldInitializeTask(AbstractMinerShieldTask):
+    def run(self, miner_shield: MinerShield):
+        # noinspection PyProtectedMember
+        miner_shield._handle_initialize()
+
+
+class MinerShieldDisableTask(AbstractMinerShieldTask):
```
Comment (andreea-popescu-reef) [inline bt_ddos_shield/miner_shield.py:445]: insteaf of define these as classes - the functions could just be passed directly

File: bt_ddos_shield/miner_shield.py [hunk 17]
```diff
@@ -1,74 +1,481 @@
...
+    def _add_task(self, task: 'AbstractMinerShieldTask'):
+        """
+        Add task to task queue. It will be handled by _worker_function.
+        """
+        if not self.run:
+            raise MinerShieldDisabledException()
+
+        self.task_queue.put(task)
+
+    def _worker_function(self):
```
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/miner_shield.py:158]: for example `def _add_task(self, task: 'AbstractMinerShieldTask'):` could mostly be replaced with `await task`

File: bt_ddos_shield/state_manager.py [hunk 18]
```diff
@@ -1,78 +1,473 @@
...
+    _validators_addresses: dict[Hotkey, Address]
+    _manifest_address: Optional[Address]
+    _address_manager_state: dict[str, str]
+    _address_manager_created_objects: dict[str, frozenset[str]]
 
+    def __init__(self, known_validators: dict[Hotkey, PublicKey], banned_validators: dict[Hotkey, datetime],
+                 validators_addresses: dict[Hotkey, Address], manifest_address: Optional[Address],
+                 address_manager_state: dict[str, str],
+                 address_manager_created_objects: dict[str, frozenset[str]]):
+        super().__setattr__('_known_validators', known_validators)
```
Comment (andreea-popescu-reef) [inline bt_ddos_shield/state_manager.py:27]: why super when there's no superclass?

File: bt_ddos_shield/state_manager.py [hunk 19]
```diff
@@ -1,78 +1,473 @@
...
+                      validators_addresses: Optional[dict[Hotkey, Address]] = None,
+                      manifest_address: Optional[Address] = None,
+                      address_manager_state: Optional[dict[str, str]] = None,
+                      address_manager_created_objects: Optional[dict[str, frozenset[str]]] = None):
+        """
+        Create new updated state with given new values and set it as current state. If value for field is None,
+        it is copied from current state.
+        """
+        self.current_miner_shield_state = \
+            MinerShieldState(dict(self.current_miner_shield_state.known_validators)
```
Comment (andreea-popescu-reef) [inline bt_ddos_shield/state_manager.py:170]: use or
```
self.current_miner_shield_state = MinerShieldState(
    known_validators or dict(self.current_miner_shield_state.known_validators),
    banned_validators or dict(self.current_miner_shield_state.banned_validators),
    validators_addresses or dict(self.current_miner_shield_state.validators_addresses),
    manifest_address if manifest_address is not None else self.current_miner_shield_state.manifest_address,
    address_manager_state or dict(self.current_miner_shield_state.address_manager_state),
    address_manager_created_objects or dict(self.current_miner_shield_state.address_manager_created_objects),
)
```

File: tests/test_address_manager.py [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+            assert AwsObjectTypes.SECURITY_GROUP.value not in created_objects
+            assert AwsObjectTypes.DNS_ENTRY.value not in created_objects
+        finally:
+            address_manager.clean_all()
+
+    def test_handle_address(self):
+        """ Create address, validate it and remove created address. """
+        address_manager: AwsAddressManager = self.create_aws_address_manager()
+
+        try:
```
Comment (andreea-popescu-reef) [inline tests/test_address_manager.py:113]: try to use https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended for the address_manager creation & cleanuup

**Comment by zyzniewski-reef** [comment]
[Bittensor BlockchainManager using commitments of knowledge](https://github.com/ggleszczynski/bt-ddos-shield/pull/1)

File: bt_ddos_shield/manifest_manager.py [hunk 21]
```diff
@@ -1,36 +1,254 @@
...
+    def _custom_decoder(json_mapping: dict[str, Any]) -> Any:
+        if "encrypted_address_mapping" in json_mapping:
+            decoded_mapping: dict[Hotkey, bytes] = {}
+            for hotkey, encoded_address in json_mapping["encrypted_address_mapping"].items():
+                decoded_mapping[Hotkey(hotkey)] = base64.b64decode(encoded_address.encode())
+            json_mapping["encrypted_address_mapping"] = decoded_mapping
+        return json_mapping
 
 
 class AbstractManifestManager(ABC):
```
[the remark this answers is not in the document]
Comment (grzegorz-leszczynski-reef) [reply bt_ddos_shield/manifest_manager.py:115]: Hmm, I wanted to have serializer objects to allow different serializations (if needed) for Manifest and Address classes. If serialization would be inside these classes, then I need to add abstract serialize/deserialize methods there and implement it in subclasses (or eventually have default serialize/deserialize in these base classes). Anyway this needs subclassing of simple Address/Manifest dataclass and I think pattern with serializer class looks nicer, but maybe I'm wrong. Or maybe you just want to remove abstract serialization?

Another thing is, if I will remove these serializer objects, then I should pass Address/Manifest class object in constructors of AbstractBlockchainManager/AbstractManifestManager to allow calling deserialize method from proper class? Or pass there only deserialize method? For serialization it is easy, because we have instance of class to call serialize on it, but when calling deserialize, we need to know which method to use.

File: bt_ddos_shield/manifest_manager.py [hunk 22]
```diff
@@ -1,36 +1,254 @@
...
+    Class representing manifest file containing encrypted addresses for validators.
+    """
+
+    encrypted_address_mapping: dict[Hotkey, bytes]
+    """ Mapping with addresses for validators (validator HotKey -> encrypted address) """
+    md5_hash: str
+    """ MD5 hash of the manifest data """
+
+
+class AbstractManifestSerializer(ABC):
```
[the remark this answers is not in the document]
Comment (grzegorz-leszczynski-reef) [reply bt_ddos_shield/manifest_manager.py:47]: I'm waiting here for response if serializers are needed at all. But if it should be combined, then I should add this abstract to utils.py file?

File: bt_ddos_shield/manifest_manager.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (grzegorz-leszczynski-reef) [reply bt_ddos_shield/manifest_manager.py:115]: I thought about asking Github Copilot about it and got such response:

Q: Is it better to have serializer class or have deserialize and serialize methods inside serialized object?

A: It is generally better to have a separate serializer class rather than having deserialize and serialize methods inside the serialized object. This approach follows the Single Responsibility Principle, making the code more modular, easier to maintain, and test. It also allows for different serialization strategies without modifying the serialized object.

Here is a brief comparison:

Separate Serializer Class

Pros:  
Adheres to the Single Responsibility Principle.
Easier to maintain and test.
Allows for different serialization strategies.
Keeps the serialized object focused on its core responsibilities.
Cons:  
Requires additional classes and interfaces.

Methods Inside Serialized Object

Pros:  
Simpler structure with fewer classes.
Serialization logic is directly associated with the object.
Cons:  
Violates the Single Responsibility Principle.
Harder to maintain and test.
Difficult to change serialization strategy without modifying the object.

In summary, using a separate serializer class is a more flexible and maintainable approach.

File: tests/test_address_manager.py (hunk 20 above)
Comment (ggleszczynski) [reply tests/test_address_manager.py]: Ok

File: bt_ddos_shield/state_manager.py (hunk 19 above)
Comment (ggleszczynski) [reply bt_ddos_shield/state_manager.py:170]: These dict objects can be empty or None and we need to treat this differently. I changed to 'or' only for manifest_address.

File: bt_ddos_shield/state_manager.py (hunk 18 above)
Comment (ggleszczynski) [reply bt_ddos_shield/state_manager.py:27]: I made this class immutable by overloading __setattr__ method. That's why I need to call setattr from super class to set these attributes. I asked copilot about this and got response, that this initialization can be done other way: `self.__dict__['_known_validators'] = known_validators`. I'm not sure if it is better than calling setattr from super class. Tell me :-)

File: bt_ddos_shield/miner_shield.py (hunk 16 above)
Comment (ggleszczynski) [reply bt_ddos_shield/miner_shield.py:445]: Hmm, I'm not sure about this. Adding these tasks as class allows us to print their names in logs when handling - I know I can use method.__name__ for this but getting class name looks nicer for me. Also some tasks have params, so I will need to pass method and params to queue, so anyway I need to wrap them in some tuple or something - having class for this also looks nicer for me.

File: bt_ddos_shield/miner_shield.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/miner_shield.py:158]: I'm not sure if it can be done with asyncio tasks, because I don't wait here for the result and task is retried forever by worker until shield gets stopped (this behaviour is by design). I will learn more about asyncio and then think if it can be changed, because my knowledge about asyncio is very small now. I wrote it in my notes for future upgrade but as for now I think this can be left as it is.

File: bt_ddos_shield/manifest_manager.py [hunk 23]
```diff
@@ -1,36 +1,254 @@
...
+        try:
+            response = self.s3_client.get_object(Bucket=self.bucket_name, Key=address.address_id)
+            return response['Body'].read()
+        except ClientError as e:
+            if e.response['Error']['Code'] == "NoSuchKey":
+                raise ManifestNotFoundException(f"File {address.address_id} not found")
+            raise e
+
+    @classmethod
+    def _serialize_manifest_address(cls, region_name: str, bucket_name: str, file_key: str) -> str:
```
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/manifest_manager.py:231]: Indeed, but here is concrete implementation (S3ManifestManager) and there was just some abstract object, which can be used in many different places. That's why I decided there to use abstract serializer and not use it here to not complicate it when not needed. See my comments for your comment there.

File: bt_ddos_shield/manifest_manager.py [hunk 24]
```diff
@@ -1,36 +1,254 @@
...
+        super().__init__(address_serializer, manifest_serializer, encryption_manager)
+        self.region_name = region_name
+        self.bucket_name = bucket_name
+        if region_name is not None:
+            self.s3_client = boto3.client("s3", aws_access_key_id=aws_access_key_id,
+                                          aws_secret_access_key=aws_secret_access_key, region_name=region_name)
+
+    def create_client_from_address(self, address: Address, aws_access_key_id: str, aws_secret_access_key: str):
+        region_name, bucket_name, _ = self._deserialize_manifest_address(address)
+        self.s3_client = boto3.client("s3", aws_access_key_id=aws_access_key_id,
```
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/manifest_manager.py]: Done

File: bt_ddos_shield/miner_shield.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/miner_shield.py:158]: I read something, thought about it, used AI and probably this can be refactored. But when current behavior is to be maintained, using asyncio won't change too much - I can use asyncio.to_thread to run _worker_function, use asyncio.Queue() instead of just Queue and make rest of needed changes, but logic and code structure will looks nearly the same (if I'm thinking correctly). So my question is, if it is worth time to change it, if it already works? I don't know the answer... Next time, when doing similar job, I'll try to use asyncio, but should I spend time for changing it here?

File: bt_ddos_shield/manifest_manager.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/manifest_manager.py:52]: yes probably

File: bt_ddos_shield/state_manager.py (hunk 18 above)
Comment (andreea-popescu-reef) [reply bt_ddos_shield/state_manager.py:37]: fair

File: bt_ddos_shield/manifest_manager.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/manifest_manager.py:120]: yea, I understand why you wrote it that way - I'm just saying that for a project of this scope having the whole logic for an abstract serializer when just one serialization/deserialization is implemented it's a bit overkill as it can just be added when needed. - as I said this is more of a code style observation

File: bt_ddos_shield/miner_shield.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/miner_shield.py:174]: I'd say this is worth refactoring - there's no point in reimplementing the logic of a lib so common as asyncio. (https://docs.python.org/3/library/asyncio-task.html#awaitables)
The refactor wouldn't take long and it'll make the code much easier to follow for any py dev.

File: bt_ddos_shield/miner_shield.py (hunk 16 above)
Comment (andreea-popescu-reef) [reply bt_ddos_shield/miner_shield.py:460]: This goes together with the asyncio refactor suggested above - in a standard asyncio app you'd expect this to be passed as function (The tuple you mentioned translates to a closure)
Having a class for this looks nice from an OOP pov but it's unnatural in python.

File: bt_ddos_shield/manifest_manager.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ggleszczynski) [reply bt_ddos_shield/manifest_manager.py:120]: Yeah, but Paweł wanted this to be as much customizable as it can be. He said that it will be used later by other users, in other clouds than AWS, etc., so that's why I've already made it abstract.

File: bt_ddos_shield/miner_shield.py (hunk 16 above)
Comment (ggleszczynski) [reply bt_ddos_shield/miner_shield.py:460]: Ok, I'll change it too

File: bt_ddos_shield/manifest_manager.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/manifest_manager.py:120]: fair

File: bt_ddos_shield/miner_shield.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/miner_shield.py:174]: see note in https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread about the GIL - I'd rather use `asyncio.create_task` instead of `to_thread` since there's no need for multiple threads. For the "io heavy" db work there is optimized native asyncio support in sqlalchemy https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html.
the Queue is actually the event loop from asyncio - it shouldn't be needed

[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply bt_ddos_shield/miner_shield.py:174]: although you might want to keep the queue just for preventing gc of the task... see `Important:` https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

**Comment by grzegorz-leszczynski-reef** [comment]
Merging PR after consultations with Even. More changes will be made.

File: bt_ddos_shield/miner_shield.py [hunk 25]
```diff
@@ -1,74 +1,631 @@
...
+        """
+        Function called in separate thread by enable() to start the shield. It is handling events put to task_queue.
+        """
+        self._event("Starting shield")
+
+        while self.run:
+            task: AbstractMinerShieldTask = self.task_queue.get()
```
Comment (zyzniewski-reef) [inline bt_ddos_shield/miner_shield.py:181]: what if a worker waits for a task (queue is empty) and you decide to disable the shield?
